hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -0,0 +1,1504 @@
+          
+          var curSym = TempSymbol(None, "tmp")
+          val inst = Instantiate(Select(c.sym.asPath, Tree.Ident("class"))(N), paramArgs)
+          var acc = blk => Assign(curSym, inst, blk)
+          for ps <- auxSyms do
+            val call = Call(curSym.asPath, ps.map(_.asPath.asArg))(true, false)
+            curSym = TempSymbol(None, "tmp")
+            acc = blk => acc(Assign(curSym, call, blk))
+          val bod = acc(Return(Call(curSym.asPath, extraSyms.map(_.asPath.asArg))(true, false), false))
+
+          inline def toPlist(ls: List[VarSymbol]) = PlainParamList(ls.map(s => Param(FldFlags.empty, s, N)))
+
+          val paramPlist = paramSyms.map(toPlist)
+          val auxPlist = auxSyms.map(toPlist)
+          val extraPlist = toPlist(extraSyms)
+
+          val plist = paramPlist match
+            case None => extraPlist :: PlainParamList(Nil) :: auxPlist
+            case Some(value) => extraPlist :: value :: auxPlist
+
+          val fakeCtorDefn = FunDefn(
+            None, bms, plist, bod
+          )
+
+          val paramSym2 = paramSyms.getOrElse(Nil)
+          val auxSym2 = auxSyms.flatMap(l => l)
+          val allSymsMp = (paramSym2 ++ auxSym2 ++ extraSyms).map(s => s -> VarSymbol(s.id)).toMap
+          val subst = new SymbolSubst():
+            override def mapVarSym(s: VarSymbol): VarSymbol = allSymsMp.get(s) match
+              case None => s
+              case Some(value) => value
+
+          val headParams = paramPlist match
+            case None => extraPlist
+            case Some(value) => ParamList(value.flags, extraPlist.params ++ value.params, value.restParam)
+
+          val auxCtorDefn_ = FunDefn(None, singleCallBms, headParams :: auxPlist, bod)
+          val auxCtorDefn = BlockTransformer(subst).applyFunDefn(auxCtorDefn_)
+          
+          Lifted(lifted, extras ::: (fakeCtorDefn :: auxCtorDefn :: Nil))
+        case c: ClsLikeDefn if modOrObj(c) => // module or object
+          // force it to be a class
+          val newK = c.k match
+            case syntax.Mod => syntax.Mod
+            case syntax.Obj => syntax.Cls
+            case _ => c.k // unreachable
+          
+          val newDef = c.copy(
+            k = newK, paramsOpt = N,
+            owner = N, auxParams = PlainParamList(extraParams) :: Nil
+          )
+          liftDefnsInCls(newDef, newCtx)
+
+        case _ => Lifted(d, Nil)
+  
+  def liftDefnsInCls(c: ClsLikeDefn, ctx: LifterCtx): Lifted[ClsLikeDefn] =
+    val ctxx = if c.k is syntax.Mod then ctx.inModule(c) else ctx
+    
+    val (preCtor, preCtorDefns) = c.preCtor.floatOut(ctxx)
+    val (ctor, ctorDefns) = c.ctor.floatOut(ctxx)
+
+    val allCtorDefns = preCtorDefns ++ ctorDefns
+    val (ctorIgnored, ctorIncluded) = allCtorDefns.partition(d => ctxx.ignored(d.sym))
+
+    // if this is a module, add a method to reference the classes lifted out
+    val extraMethods = if c.k is syntax.Mod then ctorIncluded.collect:
```

Comment:
Why is this only done if it's a module?

Is that related?
```fs
:lift
class C with
  class D()

let c = new C
//│ c = C

c.D
//│ ═══[RUNTIME ERROR] Error: Access to required field 'D' yielded 'undefined'
```

Reply:
I later decided to not lift classes inside modules.

Also, I don't know if this can be fixed with the current IR, at least without eta-expanding the constructor or making it a getter that allocates a lambda on each call.

Reply:
Nevermind, it seems you addressed this in a later comment https://github.com/hkust-taco/mlscript/pull/266#discussion_r1966445502